src/pal/src/exception/signal.cpp

```diff
@@ -240,35 +236,88 @@ void SEHCleanupSignals()
 /* internal function definitions **********************************************/
 
 #if !HAVE_MACH_EXCEPTIONS
+
 /*++
 Function :
-    sigill_handler
+    get_action_handler
 
-    handle SIGILL signal (EXCEPTION_ILLEGAL_INSTRUCTION, others?)
+    returns sa_sigaction or sa_handler depending on SA_SIGINFO flag.
 
 Parameters :
-    POSIX signal handler parameter list ("man sigaction" for details)
+    action : sigaction struct
+
+Return value :
+    handler as a SAHANDLERFUNC pointer
+--*/
+static SAHANDLERFUNC get_action_handler(struct sigaction* action)
+{
+    if (action->sa_flags & SA_SIGINFO)
+    {
+        return (SAHANDLERFUNC)action->sa_sigaction;
+    }
+    else
+    {
+        return action->sa_handler;
+    }
+}
+
+/*++
+Function :
+    invoke_previous_action
+
+    synchronously invokes the previous action or aborts when that is not possible
+
+Parameters :
+    action  : previous sigaction struct
+    code    : signal code
+    siginfo : signal siginfo
+    context : signal context
+    signalRestarts: BOOL state : TRUE if the process will be signalled again
 
     (no return value)
 --*/
-static void sigill_handler(int code, siginfo_t *siginfo, void *context)
+static void invoke_previous_action(struct sigaction* action, int code, siginfo_t *siginfo, void *context, bool signalRestarts = true)
 {
-    if (PALIsInitialized())
+    _ASSERTE(action != NULL);
+
+    SAHANDLERFUNC handler = get_action_handler(action);
```

Comment:
I would prefer not to extract getting the action handler into a separate function. This keeps the behavior that I was worried about - that we compare the handler taken from the `sa_sigaction` with the SIG_IGN / SIG_DFL. That's what I wanted to avoid. If on some Unix the SIG_IGN is defined as 0, the code below would not work as expected.
So I'd like to check for the SIG_IGN / SIG_DFL only if the `sa_handler` was the source of the handler.

Reply:
I've pushed a commit that should address this.